javascript/deep-copy.md

```diff
@@ -0,0 +1,70 @@
+# Deep Copy
+
+In programming, there are two fundamental ways of copying objects: Shallow copy and Deep copy.
+In order to understand deep copy, we need to first understand what is shallow copy.
+An Analogy: If a teacher writes some text on the blackboard and students copy the text in their notebook,
+that is deep copy, because the text is both on the blackboard and in the notebooks.
+If a teachers reads the text from a book and students only note down page numbers, then it is shallow copy,
+because the text is _only_ present in the book and everyone (including the teacher) is referring to the
+page numbers of the book.
+
+## Shallow copy
+In shallow copy, only references are copied. Both references point to the same object in the memory.
+Consider the following array of employee objects:
+```
+> const lodash = require('lodash') // JS Library that contains utility functions like clone and cloneDeep
+> employees = [
+... { "name": "John", "age": 30, salary: 5000},
+... { "name": "Ana", "age": 28, salary: 6000},
+... { "name": "Ali", age: 24, salary: 5500}
+... ]
+[
+  { name: 'John', age: 30, salary: 5000 },
+  { name: 'Ana', age: 28, salary: 6000 },
+  { name: 'Ali', age: 24, salary: 5500 }
+]
+> shallow_copy_of_employees = lodash.clone(employees) // shallow_copy_of_employees = employees //also makes shallow copy
```

Comment:
I think we can replace this with a native javascript function https://developer.mozilla.org/en-US/docs/Web/API/structuredClone

Reply:
That is quite new and only Node 17 supports it for now on backend, would be good to mention that a standard function is being developed and reference to it but that lodash is an alternative that will work in all environments whereas the standard can be used in all modern browsers.

Reply:
I will add a note, thank you @sarea  and @robvk